clients/pkg/promtail/targets/gcplog/push_translation.go

```diff
@@ -0,0 +1,117 @@
+	} `json:"message"`
+	Subscription string `json:"subscription"`
+}
+
+// translate converts a GCP PushMessage into a loki api.Entry. It is responsible for decoding the log line (contained in the Message.Data)
+// attribute, using the incoming timestamp if necessary, and formatting and passing down the incoming Message.Attributes
+// if relabel is configured.
+func translate(m PushMessage, other model.LabelSet, useIncomingTimestamp bool, relabelConfigs []*relabel.Config, xScopeOrgID string) (api.Entry, error) {
+	// mandatory label for gcplog
+	lbs := labels.NewBuilder(nil)
+	lbs.Set("__gcp_message_id", m.Message.ID)
+
+	// labels from gcp log entry. Add it as internal labels
+	for k, v := range m.Message.Attributes {
+		lbs.Set(fmt.Sprintf("__gcp_attributes_%s", convertToLokiCompatibleLabel(k)), v)
+	}
+
+	var processed labels.Labels
+
+	// apply relabeling
+	if len(relabelConfigs) > 0 {
+		processed = relabel.Process(lbs.Labels(), relabelConfigs...)
+	} else {
+		processed = lbs.Labels()
+	}
+
+	// final labelset that will be sent to loki
+	labels := make(model.LabelSet)
+	for _, lbl := range processed {
+		// ignore internal labels
+		if strings.HasPrefix(lbl.Name, "__") {
+			continue
+		}
+		// ignore invalid labels
+		if !model.LabelName(lbl.Name).IsValid() || !model.LabelValue(lbl.Value).IsValid() {
+			continue
+		}
+		labels[model.LabelName(lbl.Name)] = model.LabelValue(lbl.Value)
+	}
+
+	// add labels coming from scrapeconfig
+	labels = labels.Merge(other)
+
+	ts := time.Now()
+
+	decodedData, err := base64.StdEncoding.DecodeString(m.Message.Data)
+	if err != nil {
+		return api.Entry{}, fmt.Errorf("failed to decode data: %w", err)
+	}
+	line := string(decodedData)
+
+	if useIncomingTimestamp {
+		var err error
+		ts, err = time.Parse(time.RFC3339, m.Message.PublishTimestamp)
+		if err != nil {
+			return api.Entry{}, fmt.Errorf("invalid publish timestamp format: %w", err)
+		}
+	}
+
+	// If the incoming request carries the tenant id, inject it as the reserved label, so it's used by the
+	// remote write client.
+	if xScopeOrgID != "" {
+		labels[lokiClient.ReservedLabelTenantID] = model.LabelValue(xScopeOrgID)
+	}
+
+	return api.Entry{
+		Labels: labels,
+		Entry: logproto.Entry{
+			Timestamp: ts,
+			Line:      line,
+		},
+	}, nil
+}
+
+// convertToLokiCompatibleLabel converts an incoming GCP Push message label to a loki compatible format. There are lables
+// such as `logging.googleapis.com/timestamp`, which contain non-loki-compatible characters, which is just alphanumeric
+// and _. The approach taken is to translate every non-alphanumeric separator character to an underscore.
+func convertToLokiCompatibleLabel(label string) string {
+	// TODO: Since this is running for every incoming message, maybe it's more performant to do something
+	// like a loop over characters, and checking if it's not loki compatible, instead of a regexp.
+	return util.SnakeCase(labelToLokiCompatible.ReplaceAllString(label, "_"))
```

Comment:
Good observation about using loop instead of regexp. But let's may be keep that for later, if that end up as bottleneck.

Reply:
For sure, will leave the comment in case we need to re-evaluate

Reply:
Sorry. I must disagree. This is a rather low hanging optimization that we will probably not attend afterwards. `util.SnakeCase` is already looping through the string. You could a simple method that also escapes characters.